File: purchases/src/main/java/com/revenuecat/purchases/Backend.java
package com.revenuecat.purchases;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;

class Backend {
    final private String apiKey;
    final private Dispatcher dispatcher;
    final private HTTPClient httpClient;
    final private PurchaserInfo.Factory purchaserInfoFactory;
    final private Entitlement.Factory entitlementFactory;
    final private Map<String, String> authenticationHeaders;

    public static abstract class BackendResponseHandler {
        abstract public void onReceivePurchaserInfo(PurchaserInfo info);
        abstract public void onError(int code, String message);
    }

    public static abstract class EntitlementsResponseHandler {
        abstract public void onReceiveEntitlements(Map<String, Entitlement> entitlements);
        abstract public void onError(int code, String message);
    }

    private abstract class PurchaserInfoReceivingCall extends Dispatcher.AsyncCall {
        final private BackendResponseHandler handler;
        PurchaserInfoReceivingCall(BackendResponseHandler handler) {
            this.handler = handler;
        }

        @Override
        public void onCompletion(HTTPClient.Result result) {
            if (result.responseCode < 300) {
                try {
                    handler.onReceivePurchaserInfo(purchaserInfoFactory.build(result.body));
                } catch (JSONException e) {
                    handler.onError(result.responseCode, e.getMessage());
                }
            } else {
                String errorMessage = null;
                try {
                    String message = result.body.getString("message");
                    errorMessage = "Server error: " + message;
                } catch (JSONException jsonException) {
                    errorMessage = "Unexpected server error " + result.responseCode;
                }

                handler.onError(result.responseCode, errorMessage);
            }
        }

        @Override
        void onError(int code, String message) {
            handler.onError(code, message);
        };
    }

    Backend(String apiKey, Dispatcher dispatcher, HTTPClient httpClient, PurchaserInfo.Factory purchaserInfoFactory, Entitlement.Factory entitlementFactory) {
        this.apiKey = apiKey;
        this.dispatcher = dispatcher;
        this.httpClient = httpClient;
        this.purchaserInfoFactory = purchaserInfoFactory;
        this.entitlementFactory = entitlementFactory;

        this.authenticationHeaders = new HashMap<>();
        this.authenticationHeaders.put("Authorization", "Bearer " + this.apiKey);
    }

    public void getSubscriberInfo(final String appUserID, final BackendResponseHandler handler) {
        dispatcher.enqueue(new PurchaserInfoReceivingCall(handler) {
            @Override
            public HTTPClient.Result call() throws HTTPClient.HTTPErrorException {
                return httpClient.performRequest("/subscribers/" + appUserID, (Map)null, authenticationHeaders);
            }
        });
    }

    public void postReceiptData(final String purchaseToken, final String appUserID, final String productID, final Boolean isRestore, final BackendResponseHandler handler) {
        final Map<String, Object> body = new HashMap<>();

        body.put("fetch_token", purchaseToken);
        body.put("product_id", productID);
        body.put("app_user_id", appUserID);
        body.put("is_restore", isRestore);

        dispatcher.enqueue(new PurchaserInfoReceivingCall(handler) {
            @Override
            public HTTPClient.Result call() throws HTTPClient.HTTPErrorException {
                return httpClient.performRequest("/receipts", body, authenticationHeaders);
            }
        });
    }

    void getEntitlements(final String appUserID, final EntitlementsResponseHandler handler) {
        dispatcher.enqueue(new Dispatcher.AsyncCall() {
            @Override
            public HTTPClient.Result call() throws HTTPClient.HTTPErrorException {
                return httpClient.performRequest("/subscribers/" + appUserID + "/products",(Map)null, authenticationHeaders);
            }

            void onError(int code, String message) {
                handler.onError(code, message);
            }

            void onCompletion(HTTPClient.Result result) {
                if (result.responseCode < 300) {
                    try {
                        JSONObject entitlementsResponse = result.body.getJSONObject("entitlements");
                        Map<String, Entitlement> entitlementMap = entitlementFactory.build(entitlementsResponse);
                        handler.onReceiveEntitlements(entitlementMap);
                    } catch (JSONException e) {
                        handler.onError(result.responseCode, "Error parsing products JSON " + e.getLocalizedMessage());
                    }
                } else {
                    handler.onError(result.responseCode, "Backend error");
                }
            }
        });
    }

    void postAttributionData(final String appUserID, @Purchases.AttributionNetwork int network, JSONObject data) {
        if (data.length() == 0) return;

        final JSONObject body = new JSONObject();
        try {
            body.put("network", network);
            body.put("data", data);
        } catch (JSONException e) {
            return;
        }

        dispatcher.enqueue(new Dispatcher.AsyncCall() {
            @Override
            public HTTPClient.Result call() throws HTTPClient.HTTPErrorException {
                return httpClient.performRequest("/subscribers/" + appUserID + "/attribution", body, authenticationHeaders);
            }
        });
    }
}

File: purchases/src/main/java/com/revenuecat/purchases/Backend.kt
package com.revenuecat.purchases

import android.net.Uri

import org.json.JSONException
import org.json.JSONObject

import java.util.HashMap

internal class Backend(
    private val apiKey: String,
    private val dispatcher: Dispatcher,
    private val httpClient: HTTPClient,
    private val purchaserInfoFactory: PurchaserInfo.Factory,
    private val entitlementFactory: Entitlement.Factory
) {

    internal val authenticationHeaders: MutableMap<String, String>

    abstract class BackendResponseHandler {
        abstract fun onReceivePurchaserInfo(info: PurchaserInfo)
        abstract fun onError(code: Int, message: String?)
    }

    abstract class EntitlementsResponseHandler {
        abstract fun onReceiveEntitlements(entitlements: Map<String, Entitlement>)
        abstract fun onError(code: Int, message: String)
    }

    abstract class AliasResponseHandler {
        abstract fun onSuccess()
        abstract fun onError(code: Int, message: String)
    }

    private abstract inner class PurchaserInfoReceivingCall internal constructor(private val handler: BackendResponseHandler) :
        Dispatcher.AsyncCall() {

        public override fun onCompletion(result: HTTPClient.Result) {
            if (result.responseCode < 300) {
                try {
                    handler.onReceivePurchaserInfo(purchaserInfoFactory.build(result.body!!))
                } catch (e: JSONException) {
                    handler.onError(result.responseCode, e.message)
                }

            } else {
                var errorMessage: String? = null
                try {
                    val message = result.body!!.getString("message")
                    errorMessage = "Server error: $message"
                } catch (jsonException: JSONException) {
                    errorMessage = "Unexpected server error " + result.responseCode
                }

                handler.onError(result.responseCode, errorMessage)
            }
        }

        internal override fun onError(code: Int, message: String) {
            handler.onError(code, message)
        }
    }

    init {

        this.authenticationHeaders = HashMap()
        this.authenticationHeaders["Authorization"] = "Bearer " + this.apiKey
    }

    fun close() {
        this.dispatcher.close()
    }

    private fun enqueue(call: Dispatcher.AsyncCall) {
        if (!dispatcher.isClosed) {
            dispatcher.enqueue(call)
        }
    }

    fun getSubscriberInfo(appUserID: String, handler: BackendResponseHandler) {
        enqueue(object : PurchaserInfoReceivingCall(handler) {
            @Throws(HTTPClient.HTTPErrorException::class)
            override fun call(): HTTPClient.Result {
                return httpClient.performRequest(
                    "/subscribers/" + encode(appUserID),
                    null as Map<*, *>?,
                    authenticationHeaders
                )
            }
        })
    }

    fun postReceiptData(
        purchaseToken: String,
        appUserID: String,
        productID: String,
        isRestore: Boolean?,
        handler: BackendResponseHandler
    ) {
        val body = HashMap<String, Any?>()

        body["fetch_token"] = purchaseToken
        body["product_id"] = productID
        body["app_user_id"] = appUserID
        body["is_restore"] = isRestore

        enqueue(object : PurchaserInfoReceivingCall(handler) {
            @Throws(HTTPClient.HTTPErrorException::class)
            override fun call(): HTTPClient.Result {
                return httpClient.performRequest("/receipts", body, authenticationHeaders)
            }
        })
    }

    fun getEntitlements(appUserID: String, handler: EntitlementsResponseHandler) {
        enqueue(object : Dispatcher.AsyncCall() {
            @Throws(HTTPClient.HTTPErrorException::class)
            override fun call(): HTTPClient.Result {
                return httpClient.performRequest(
                    "/subscribers/" + encode(appUserID) + "/products",
                    null as Map<*, *>?,
                    authenticationHeaders
                )
            }

            internal override fun onError(code: Int, message: String) {
                handler.onError(code, message)
            }

            internal override fun onCompletion(result: HTTPClient.Result) {
                if (result.responseCode < 300) {
                    try {
                        val entitlementsResponse = result.body!!.getJSONObject("entitlements")
                        val entitlementMap = entitlementFactory.build(entitlementsResponse)
                        handler.onReceiveEntitlements(entitlementMap)
                    } catch (e: JSONException) {
                        handler.onError(
                            result.responseCode,
                            "Error parsing products JSON " + e.localizedMessage
                        )
                    }

                } else {
                    handler.onError(result.responseCode, "Backend error")
                }
            }
        })
    }

    private fun encode(string: String): String {
        return Uri.encode(string)
    }

    fun postAttributionData(
        appUserID: String,
        network: Purchases.AttributionNetwork,
        data: JSONObject
    ) {
        if (data.length() == 0) return

        val body = JSONObject()
        try {
            body.put("network", network.serverValue)
            body.put("data", data)
        } catch (e: JSONException) {
            return
        }

        enqueue(object : Dispatcher.AsyncCall() {
            @Throws(HTTPClient.HTTPErrorException::class)
            override fun call(): HTTPClient.Result {
                return httpClient.performRequest(
                    "/subscribers/" + encode(appUserID) + "/attribution",
                    body,
                    authenticationHeaders
                )
            }
        })
    }

    fun alias(
        appUserID: String,
        newAppUserID: String,
        onSuccessHandler: () -> Unit,
        onErrorHandler: (Int, String) -> Unit
    ) {
        val body = mapOf(
            "new_app_user_id" to newAppUserID
        )

        enqueue(object : Dispatcher.AsyncCall() {
            @Throws(HTTPClient.HTTPErrorException::class)
            override fun call(): HTTPClient.Result {
                return httpClient.performRequest(
                    "/subscribers/" + encode(appUserID) + "/alias",
                    body,
                    authenticationHeaders
                )
            }

            override fun onError(code: Int, message: String) {
                onErrorHandler(code, message)
            }

            override fun onCompletion(result: HTTPClient.Result) {
                if (result.responseCode < 300) {
                    try {
                        onSuccessHandler()
                    } catch (e: JSONException) {
                        onErrorHandler(result.responseCode, "Backend error")
                    }
                } else {
                    onErrorHandler(result.responseCode, "Backend error")
                }
            }
        })
    }
}
